applications/blogs/backend/src/main/java/pl/ds/kyanite/blogs/components/models/BlogPostTableOfContentsComponent.java

```diff
@@ -74,27 +86,71 @@ public class BlogPostTableOfContentsComponent {
 
   @PostConstruct
   private void init() {
-    this.items = new ArrayList<>();
+    prepareTitlesHierarchy();
+  }
 
+  private void prepareTitlesHierarchy() {
+    this.subTitles = new ArrayList<>();
     Resource pageResource = getContainingPageAsResource(this.resource);
 
     if (pageResource != null) {
       Resource contentsContainer = resourceResolver.getResource(
           pageResource.getPath() + BLOG_CONTENT_CONTAINER_REL_PATH);
-      List<Resource> titles = findAllResourcesInContainerByType(
-          contentsContainer, TITLE_RESOURCE_PATH);
-
-      for (Resource contentTitle : titles) {
-        TitleComponent titleComponent = contentTitle.adaptTo(TitleComponent.class);
-
-        if (!"p".equals(titleComponent.getElement())) {
-          BlogPostTableOfContentsItemComponent contentItem =
-              prepareNewContentItem(titleComponent);
-          if (contentItem != null) {
-            this.items.add(contentItem);
+      List<Resource> titles = findAllValidTitlesInContainer(contentsContainer);
```

Comment:
1. ok, makes sense as we never use Resource
2. You're right, we don't need to compare to NO_HEADING_LEVEL again. I left it here because otherwise the reader has to keep in mind that this check has already performed in other place. Now I think I'd better add this block as the 'else' block at the end.